text/0000-rewards.md

```diff
@@ -0,0 +1,250 @@
+
+We propose a simple approximate solution based upon computing medians across validators for used votes.
+
+0. In an epoch $e$, each validator $u$ counts of the number $\alpha_{u,v}$ of votes they *used* from each validator $v$, including themselves.  Any time a validator marks a candidate approved, they increment these counts appropriately. 
+
+1. After epoch $e$'s last block gets finalized, all validators of epoch $e$ submit an _approvals tally message_ `ApprovalsTallyMessage` that reveals their number $\alpha_{u,v}$ of useful approvals they saw from each validator $v$ on candidates that became available in epoch $n$.  We do not send $\alpha_{u,u}$ for tit-for-tat reasons discussed below, not for bias concerns.  We record these approvals tally messages on-chain.
+
+2. After some delay, we compute on-chain the median $\alpha_v := \textrm{median} \{ \alpha_{u,v} : u \}$ used approvals statements for each validator $v$. 
+
+As discussed in https://hackmd.io/@rgbPIkIdTwSICPuAq67Jbw/S1fHcvXSF we could compute these medians using the [on-line algorithm](https://www.quora.com/Is-there-an-online-algorithm-to-calculate-the-median-of-a-stream-of-numbers-if-stream-elements-can-be-added-or-removed-at-any-point?share=1) if substrate had a nice priority queue.
+
+We never achieve true consensus on approval checkers and their approval votes.  Yet, our approval assignment loop gives a rough concensus, under our Byzantine assumption and some synchrony assumption.  It then follows that miss-reporting by malicious validators should not appreciably alter the median $\alpha_v$ and hence rewards.  
+
+We never tally used approval assignments to candidate equivocations or other forks.  Any validator should always conclude whatever approval checks it begins, even on other forks, but we expect relay chain equivocations should be vanishingly rare, and sassafras should make forks uncommon.
+
+### Availability redistribution
+
+As approval checkers could easily perform useless checks, we shall reward availability providers for the availability chunks they provide that resulted in useful approval checks.  We enforce honesty using a tit-for-tat mechanism because chunk transfers are inherently subjective.
+
+An approval checker reconstructs the full parachain block by downloading distinct $f+1$ chunks from other validators, where at most $f$ validators are byzantine, out of the $n \ge 3 f + 1$ total validators.  In downloading chunks, validators prefer the $f+1$ systemic chunks over the non-systemic chunks, and prefer fetching from validators who already voted valid, like backing checkers.  It follows some validators should recieve credit for more than one chunk per candidate.  
```

Comment:
In practice, we almost always first try to recover the full unencoded PoV from one of the validators in the backing group. We fallback to systematic chunks and regular chunks if this fails OR if the PoV size is larger than 1 Mib.

Would the validators in the backing group be rewarded as if they had provided all of the chunks (which they did)? Or are we going to ignore these

Reply:
That's actually what this says.  Fetching from backers is one of the two ways to fetch the systemic chunks.  Systemic chunks is just means the original parablock.  It's fine if this is rephrased to better match the terminology elsewhere though.

> Would the validators in the backing group be rewarded as if they had provided all of the chunks (which they did)? Or are we going to ignore these

Yes, backers are rewarded for every chunk they give out (assuming the. guy who gets it credits them).  This is why backers are only paid like 80% of what approval checkers are paid, instead of something higher like 98% or whatever.